controller/controlplane/controller_reconciler_utils.go

```diff
@@ -348,12 +356,306 @@ func (r *Reconciler) ensureRoles(
+		if err != nil {
+			return false, err
+		}
+		if !objDeleted {
+			continue
+		}
+
+		deleted = true
+		log.Debug(logger, "Role deleted", "role", role)
+	}
+
+	return deleted, nil
+}
+
+func (r *Reconciler) pruneOutdatedRoleBindings(
+	ctx context.Context,
+	logger logr.Logger,
+	cp *operatorv1beta1.ControlPlane,
+	validatedWatchNamespaces []string,
+) (deleted bool, err error) {
+	namespaces := sets.New(validatedWatchNamespaces...)
+
+	roleBindings, err := k8sutils.ListRoleBindings(
+		ctx,
+		r.Client,
+		client.MatchingLabels(k8sutils.GetManagedByLabelSet(cp)),
+	)
+	if err != nil {
+		return false, err
+	}
+
+	for _, rb := range roleBindings {
+		objDeleted, err := deleteIfNotInNamespaceSet(ctx, r.Client, &rb, namespaces, "ControlPlane's RoleBinding")
+		if err != nil {
+			return false, err
+		}
+		if !objDeleted {
+			continue
+		}
+
+		deleted = true
+		log.Debug(logger, "RoleBinding deleted", "role_binding", rb)
+	}
+
+	return deleted, nil
+}
+
+func deleteIfNotInNamespaceSet(
+	ctx context.Context, cl client.Client, obj client.Object, namespaces sets.Set[string], msg string,
+) (bool, error) {
+	if namespaces.Has(obj.GetNamespace()) {
+		return false, nil
+	}
+
+	if err := cl.Delete(ctx, obj); err != nil {
+		return false, fmt.Errorf("failed deleting %s %s: %w", msg, client.ObjectKeyFromObject(obj), err)
+	}
+	return true, nil
+}
+
+func (r *Reconciler) ensureRoles(
+	ctx context.Context,
+	cp *operatorv1beta1.ControlPlane,
+	generatedRoles []*rbacv1.Role,
+) (createdOrUpdated bool, roles []*rbacv1.Role, err error) {
+rolesLoop:
+	for _, generatedRole := range generatedRoles {
+		existingRoles, err := k8sutils.ListRoles(
+			ctx,
+			r.Client,
+			client.MatchingLabels(k8sutils.GetManagedByLabelSet(cp)),
+			client.InNamespace(generatedRole.Namespace),
+		)
+		if err != nil {
+			return false, nil, err
+		}
+
+		count := len(existingRoles)
+		switch count {
+		case 0:
+			if err := r.Create(ctx, generatedRole); err != nil {
+				return false, nil, err
+			}
+			roles = append(roles, generatedRole)
+			createdOrUpdated = true
+		case 1:
+			var (
+				updated  bool
+				existing = &existingRoles[0]
+				old      = existing.DeepCopy()
+			)
+
+			updated, existing.ObjectMeta = k8sutils.EnsureObjectMetaIsUpdated(existing.ObjectMeta, generatedRole.ObjectMeta)
+			if updated ||
+				!cmp.Equal(existing.Rules, generatedRole.Rules) {
+				existing.Rules = generatedRole.Rules
+				if err := r.Patch(ctx, existing, client.MergeFrom(old)); err != nil {
+					return false, nil, fmt.Errorf("failed patching ControlPlane's Role %s: %w", existing.Name, err)
+				}
+				roles = append(roles, existing)
+				createdOrUpdated = true
+				continue rolesLoop
+			}
+			roles = append(roles, existing)
+		default:
+			if err := k8sreduce.ReduceRoles(ctx, r.Client, existingRoles); err != nil {
+				return false, nil, err
+			}
+			return false, nil, errors.New("number of Roles reduced")
```

Comment:
```suggestion
			return false, nil, fmt.Errorf("number of Roles reduced from: %d to 1", count)
```

a little bit more of extra context may be beneficial I think